backends/tfhe-cuda-backend/cuda/src/integer/multiplication.cuh

```diff
@@ -347,114 +363,78 @@ __host__ void host_integer_mult_radix_kb(
+  size_t h_lwe_idx_out[r * num_blocks];
+  int32_t h_smart_copy_in[r * num_blocks];
+  int32_t h_smart_copy_out[r * num_blocks];
+
   auto max_shared_memory = cuda_get_max_shared_memory(stream->gpu_index);
-  while (r > chunk_size) {
-    int cur_total_blocks = r * num_blocks;
+
+  lwe_indexes_in = luts_message_carry->lwe_indexes_in;
+  lwe_indexes_out = luts_message_carry->lwe_indexes_out;
+
+  while (r > 2) {
+    size_t cur_total_blocks = r * num_blocks;
     ch_amount = r / chunk_size;
     dim3 add_grid(ch_amount, num_blocks, 1);
     size_t sm_size = big_lwe_size * sizeof(Torus);
-    cuda_memset_async(new_blocks, 0,
-                      ch_amount * num_blocks * big_lwe_size * sizeof(Torus),
-                      stream);
 
     tree_add_chunks<Torus, params><<<add_grid, 256, sm_size, stream->stream>>>(
-        new_blocks, old_blocks, chunk_size, num_blocks);
-
-    for (int c_id = 0; c_id < ch_amount; c_id++) {
-      auto cur_chunk = &terms_degree[c_id * chunk_size * num_blocks];
-      int mx = 0;
-      int mn = num_blocks;
-      for (int r_id = 1; r_id < chunk_size; r_id++) {
-        auto cur_radix = &cur_chunk[r_id * num_blocks];
-        for (int i = 0; i < num_blocks; i++) {
-          if (cur_radix[i]) {
-            mn = min(mn, i);
-            mx = max(mx, i);
-          }
-        }
-      }
-      f_b[c_id] = mn;
-      l_b[c_id] = mx;
-    }
-
-    int total_copied = 0;
-    int message_count = 0;
-    int carry_count = 0;
-    compress_device_array_with_map<Torus>(stream, new_blocks, old_blocks, f_b,
-                                          l_b, num_blocks, ch_amount,
-                                          big_lwe_size, total_copied, true);
-
-    message_count = total_copied;
-    compress_device_array_with_map<Torus>(stream, new_blocks, old_blocks, f_b,
-                                          l_b, num_blocks, ch_amount,
-                                          big_lwe_size, total_copied, false);
-    carry_count = total_copied - message_count;
-
-    auto message_blocks_vector = old_blocks;
-    auto carry_blocks_vector =
-        &old_blocks[message_count * (glwe_dimension * polynomial_size + 1)];
+        new_blocks, old_blocks, min(r, chunk_size), num_blocks);
+
+    size_t total_count = 0;
+    size_t message_count = 0;
+    size_t carry_count = 0;
+    size_t sm_copy_count = 0;
+
+    generate_ids_update_degrees(
+        terms_degree, h_lwe_idx_in, h_lwe_idx_out, h_smart_copy_in,
+        h_smart_copy_out, ch_amount, r, num_blocks, chunk_size, message_max,
+        total_count, message_count, carry_count, sm_copy_count);
+
+    cuda_set_value_async<Torus>(
+        &(stream->stream), luts_message_carry->get_lut_indexes(message_count),
+        1, carry_count);
+    size_t copy_size = total_count * sizeof(Torus);
+    cuda_memcpy_async_to_gpu(lwe_indexes_in, h_lwe_idx_in, copy_size, stream);
+    cuda_memcpy_async_to_gpu(lwe_indexes_out, h_lwe_idx_out, copy_size, stream);
```

Comment:
Why do you have to copy the lut indexes again?

Reply:
in every while loop iteration we have different values for indexes